lib/events/events.go

```diff
@@ -351,3 +353,54 @@ func Error(err error) *string {
 	str := err.Error()
 	return &str
 }
+
+// Wait for the given event to become available, or for a timeout, whichever is sooner.
+// Returns true if the event became available, or false if we timed out.
+func (s *bufferedSubscription) waitForNewEventWithTimeout(id int, timeout time.Duration) bool {
+	// Checks whether 'id' is available. If it is, returns nil. If it isn't, returns newEventNotification (constructing it if necessary)
+	checkIDAndConstructNotificationIfNecessary := func() <-chan struct{} {
+		s.mut.Lock()
+		defer s.mut.Unlock()
+
+		if id < s.cur {
+			return nil
+		}
+		if s.newEventNotification == nil {
+			s.newEventNotification = make(chan struct{})
+		}
+		return s.newEventNotification
+	}
+
+	// Early exit, in the case where events are already available - saves creating the timer
+	newEventNotification := checkIDAndConstructNotificationIfNecessary()
+	if newEventNotification == nil {
+		return true
+	}
+	if timeout <= 0 {
+		return false
+	}
+
+	// From the point on, we know we're going to have to wait
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		// If the timer fires, then we've definitely timed out.
+		// If the newEventNotification channel is closed, then the event we want may or may not be present.
+		// It's likely that it will be, because people don't tend to ask for events in the future,
```

Comment:
Right, added a commit which attempts to do this. It gets messy as the timer needs to live through different calls to Wait(), but it's not too bad